docs/arch/adr-010-new-versioning-scheme.md

```diff
@@ -0,0 +1,193 @@
+```yaml
+apiVersion: "cluster.kurl.sh/v1beta2"
+kind: "Installer"
+metadata:
+  name: "cluster"
+spec:
+  channel:  "1.27"
+```
+_We have decided to transition to the newer **v1beta2** variant. Existing users who prefer to continue working with the outdated format may still do so at their own discretion_
+
+#### Pinning the kURL version
+
+Another feature we aim to offer is the ability for users to fix (pin) their installation to a particular kURL installer version. The organisation of kURL installer releases follows the pattern `YYYY.MM.DD-x`. A complete list of kURL versions can be seen [here](https://github.com/replicatedhq/kURL/releases).
+
+For those seeking to pin their setup to a designated kURL installer version, the `installerVersion` field must be appropriately configured as shown in the example below.
+
+```yaml
+apiVersion: "cluster.kurl.sh/v1beta2"
+kind: "Installer"
+metadata:
+  name: "cluster"
+spec:
+  channel:  "1.27"
+  installerVersion: "v2023.04.24-0"
+```
+
+While it's certainly possible to omit the `installerVersion` configuration, doing so means relying upon the current most up-to-dated kURL installer version.
+
+### Add-on versions
+
+In order to maintain greater control over the installation process, we have decided to modify our approach to add-on selection. Going forward, we will no longer allow users to freely select any add-on on any version of their choosing. Instead, we will provide users with specific bundles of add-ons (from here on this ADR will call them _Bundles_), each of which will be available in different versions within a designated Channel.
+
+This is an example of a Bundle on its v1.0.0 version:
+
+| Add-on     | Version              |
+|------------|----------------------|
+| Containerd | 1.6.20               |
+| Flannel    | 0.21.5               |
+| OpenEBS    | 3.5.0.               |
+| Minio      | 2023-03-13T19-46-17Z |
+
+
+To facilitate the selection of a Bundle within a Channel, we will implement a new feature that enables users to pin a specific version of a Bundle. This means that users can select and install the precise version of a Bundle that best aligns with their needs, providing them with reproducibility.
+
+To illustrate the concept of Bundle version within a Channel, we have provided an example of one that could be available within the Channel v1.24. Let's call this _Bundle v1.0.0_:
+
+| Add-on     | Version              |
+|------------|----------------------|
+| Containerd | 1.6.16               |
+| Flannel    | 0.21.4               |
+| OpenEBS    | 3.3.0.               |
+| Minio      | 2023-03-13T19-46-17Z |
+
+For users who wish to pin their Bundle version within a Channel, we will implemented a straightforward solution. By providing the `bundleVersion` parameter as follows, users can easily specify the exact version of the add-on Bundle that they wish to install:
```

Comment:
Why do we provide both `bundleVersion` and `installerVersion`?

Why not just tie the addon versions to the installer?

Reply:
I agree with Andrew here. Let's tie add-on versions to installer versions. Introducing the additional bundle version only adds more unnecessary variability and complexity.

Reply:
These is a list of consequences of this change, I want to collect your input about them:

1) We are choosing to deliberately change the current `installerVersion` semantic.
2) We gonna have to choose if the `installerVersion` X under 1.24 installs the same add-ons as the same `installerVersion` X under 1.27 (of if the bundled add-ons change following the Kubernetes channel).
3) We will need to "block" certain `installerVersions` to be used in very old channels.

Reply:
I have updated the ADR to remove any reference to "Bundle", please take a look.

Reply:
> 1. We are choosing to deliberately change the current installerVersion semantic.

we are using the same versioning as we have today in kurl

> 2. We gonna have to choose if the installerVersion X under 1.24 installs the same add-ons as the same installerVersion X under 1.27 (or if the bundled add-ons change following the Kubernetes channel).

I think we would like to leave ourselves open to the add-on versions not always being the same as we do not want a release of one kuberentes version to block a release of the others say if the upstream add-on did not yet support a version.

> 3. We will need to "block" newer installerVersions to be used in very old channels.

I think we have decided on the following rules to prevent an infinite matrix of test scenarios:

1. You can never downgrade the installerVersion or Kubernetes.
2. You can never move both the installerVersion and the Kubernetes minor version at the same time (You can either keep the installerVersion the same, and move to the next Kubernetes minor version, or keep the Kubernetes minor version the same and move to any newer installerVersion).
3. Once a Kubernetes version is EOL by the upstream, we will no longer release new installerVersions with it.

Reply:
> 3. We will need to "block" newer installerVersions to be used in very old channels.

we already do this today - if you try to install k8s 1.26 with an installerVersion from 2022, the API will return an error

Reply:
> > 1. We are choosing to deliberately change the current installerVersion semantic.
> 
> we are using the same versioning as we have today in kurl

Yes, the same versioning but with a different meaning. Installer versions, as they are today, do not include add-ons.

> > 2. We gonna have to choose if the installerVersion X under 1.24 installs the same add-ons as the same installerVersion X under 1.27 (or if the bundled add-ons change following the Kubernetes channel).
> 
> I think we would like to leave ourselves open to the add-on versions not always being the same as we do not want a release of one kuberentes version to block a release of the others say if the upstream add-on did not yet support a version.

Let me see if I understood you: version `v2023.04.24-0` when installed on Kubernetes 1.24 may bring different add-ons than the same version but installed on Kubernetes 1.25 ?

> > 3. We will need to "block" newer installerVersions to be used in very old channels.
> 
> I think we have decided on the following rules to prevent an infinite matrix of test scenarios:
> 
> 1. You can never downgrade the installerVersion or Kubernetes.
> 2. You can never move both the installerVersion and the Kubernetes minor version at the same time (You can either keep the installerVersion the same, and move to the next Kubernetes minor version, or keep the Kubernetes minor version the same and move to any newer installerVersion).
> 3. You can only move 1 Kubernetes minor version at a time.
> 4. Once a Kubernetes version is EOL by the upstream, we will no longer release new installerVersions with it.

Could you point me to where this has been decided ? Is this part of a document somewhere ? I would like to include this here.

Reply:
Sorry I would have been more clear about what I meant when i said "we have decided". We discussed this ADR on our weekly internal meeting and I am only summarizing the decision. You can find the notes in 4/27/2023. 

> Let me see if I understood you: version v2023.04.24-0 when installed on Kubernetes 1.24 may bring different add-ons than the same version but installed on Kubernetes 1.25 ?

the hope is that all version are the same but we may have to leave ourselves open to this possibility in order to prevent upstream projects from delaying us from shipping. this is only hypothetical and we could make this a rule until we are forced to break it.

Reply:
This has already been addressed on the ADR. Let me know if that is OK or if there is something else that needs to be added.

Reply:
LGTM as is